bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/ItemChannelLinkRegistry.java

```diff
@@ -23,65 +28,118 @@
-            ((ManagedItemChannelLinkProvider) this.managedProvider).removeLinksForThing(thingUID);
-        } else {
-            throw new IllegalStateException("ManagedProvider is not available");
-        }
-    }
+public class ItemChannelLinkRegistry extends
+		AbstractLinkRegistry<ItemChannelLink> implements ItemRegistryChangeListener {
+
+	private ThingRegistry thingRegistry;
+	private EventPublisher eventPublisher;
+	private ItemRegistry itemRegistry;
+
+	/**
+	 * Returns a set of bound channels for the given item name.
+	 *
+	 * @param itemName
+	 *            item name
+	 * @return set of bound channels for the given item name
+	 */
+	public Set<ChannelUID> getBoundChannels(String itemName) {
+
+		Set<ChannelUID> channelUIDs = new HashSet<>();
+
+		for (ItemChannelLink itemChannelLink : getAll()) {
+			if (itemChannelLink.getItemName().equals(itemName)) {
+				channelUIDs.add(itemChannelLink.getUID());
+			}
+		}
+
+		return channelUIDs;
+	}
+
+	/**
+	 * Returns a set of bound things for the given item name.
+	 *
+	 * @param itemName
+	 *            item name
+	 * @return set of bound things for the given item name
+	 */
+	public Set<Thing> getBoundThings(String itemName) {
+
+		Set<Thing> things = new HashSet<>();
+		Collection<ChannelUID> boundChannels = getBoundChannels(itemName);
+
+		for (ChannelUID channelUID : boundChannels) {
+			Thing thing = thingRegistry.get(channelUID.getThingUID());
+			if (thing != null) {
+				things.add(thing);
+			}
+		}
+
+		return things;
+	}
+
+	protected void setThingRegistry(ThingRegistry thingRegistry) {
+		this.thingRegistry = thingRegistry;
+	}
+
+	protected void unsetThingRegistry(ThingRegistry thingRegistry) {
+		this.thingRegistry = null;
+	}
+
+	protected void setItemRegistry(ItemRegistry itemRegistry) {
+		this.itemRegistry = itemRegistry;
+		itemRegistry.addRegistryChangeListener(this);
+	}
+
+	protected void unsetItemRegistry(ItemRegistry itemRegistry) {
+		this.itemRegistry = null;
+	}
+
+	protected void setEventPublisher(EventPublisher eventPublisher) {
+		this.eventPublisher = eventPublisher;
+	}
+
+	protected void unsetEventPublisher(EventPublisher eventPublisher) {
+		this.eventPublisher = null;
+	}
+
+	public void removeLinksForThing(ThingUID thingUID) {
+		if (this.managedProvider != null) {
+			((ManagedItemChannelLinkProvider) this.managedProvider)
+					.removeLinksForThing(thingUID);
+		} else {
+			throw new IllegalStateException("ManagedProvider is not available");
+		}
+	}
+
+	@Override
+	protected void onAddElement(ItemChannelLink element)
+			throws IllegalArgumentException {
+		super.onAddElement(element);
+		String itemName = element.getItemName();
+		if (itemRegistry.get(itemName) != null) {
+			eventPublisher.postCommand(itemName, RefreshType.REFRESH);
+		}
+	}
+
+	@Override
+	public void added(Item item) {
+		if(!getBoundChannels(item.getName()).isEmpty()) {
+			eventPublisher.postCommand(item.getName(), RefreshType.REFRESH);
```

Comment:
no - REFRESH is only sent once both are available. The first does not send, if the item is not yet there, the second does not send, if the link is not yet there.